SDWebImage/SDWebImageDownloaderOperation.m

```diff
@@ -103,10 +103,15 @@ - (nullable id)addHandlersForProgress:(nullable SDWebImageDownloaderProgressBloc
 
 - (nullable NSArray<id> *)callbacksForKey:(NSString *)key {
     LOCK(self.callbacksLock);
-    NSMutableArray<id> *callbacks = [[self.callbackBlocks valueForKey:key] mutableCopy];
+    NSArray<SDCallbacksDictionary *> *callbackBlocks = [self.callbackBlocks copy];
     UNLOCK(self.callbacksLock);
-    // We need to remove [NSNull null] because there might not always be a progress block for each callback
-    [callbacks removeObjectIdenticalTo:[NSNull null]];
```

Comment:
Actually....I don't know why previous writter use a `NSDictionary` to store just two type of completion block. Why not create a simple Object instead ? Without that tricky code anymore.

```objectivec
@interface SDWebImageDownloaderOperationCallback : NSObject

@property (nonatomic, copy, nullable) SDWebImageDownloaderProgressBlock progressBlock;
@property (nonatomic, copy, nullable) SDWebImageDownloaderCompletedBlock completionBlock;

@end

// Then just upgradte all internal APIs.

@property (strong, nonatomic, nonnull) NSMutableArray<SDWebImageDownloaderOperationCallback> *callbackBlocks;

// - (nullable NSArray<id> *)callbacksForKey:(NSString *)key -> Use this instead. Simpler
- (NSArray<SDWebImageDownloaderProgressBlock> *)progressBlocks {
    LOCK(self.callbacksLock);
    NSMutableArray<SDWebImageDownloaderOperationCallback *> *callbackBlocks = [self.callbackBlocks copy];
    UNLOCK(self.callbacksLock);
    NSMutableArray<SDWebImageDownloaderProgressBlock> *mutableProgressBlocks = [NSMutableArray arrayWithCapacity:callbackBlocks.count];
    for (SDWebImageDownloaderOperationCallback callbackBlock in callbackBlocks) {
        [mutableProgressBlocks addObject:callbackBlock];
    }
    return [mutableProgressBlocks copy];
}
```

Reply:
@dreampiggy , the `NSNull` check is the bug fix (not realized that would return `NSNull` ), so I think the previous writer just don't want to create a new class,it's unnecessary, just only use the `KVC` of the `NSArray` / `NSDictionary` can handle these things gracefully.